### PR TITLE
add support for custom Attributess in KeyAttributes

### PR DESCRIPTION
**Pull request by jafu888**
This adds support for custom attributes in KeyAttributes:
```json5
KeyAttributes: [
                        {
                          target: ['a'],
                          frames: [25,50,75],
                          rotationX: [0, 45, 60],
                          rotationY: [60, 45, 0],
                          custom: {
                            background: ['#FFFFFF', '#000000','#FFFFFF'],
                            count: [50,50,50]
                          }
                          
                        }
                      ]
```

https://user-images.githubusercontent.com/15019413/190953090-d2dd7417-67a7-4e83-9d9a-2b6886f13be7.mp4

### CODE REVIEW

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 1]
```diff
@@ -660,6 +661,16 @@ public void addKeyAttribute(String target, TypedBundle bundle) {
         getWidgetState(target, null, 0).setKeyAttribute(bundle);
     }
 
+    /**
+     * Add a key attribute and the custom variables into the
+     * @param target the id of the target
+     * @param bundle the key attributes bundle containing position etc.
+     * @param custom the customVariables to add at that position
+     */
+    public void addKeyAttribute(String target, TypedBundle bundle, CustomVariable[]custom) {
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:670]: nit: space before custom

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 2]
```diff
@@ -898,6 +909,23 @@ public void setKeyAttribute(TypedBundle prop) {
             mMotionControl.addKey(keyAttributes);
         }
 
+        /**
+         * Set tge keyAttribute bundle and associated custom attributes
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:913]: nit: the

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 3]
```diff
@@ -660,6 +661,16 @@ public void addKeyAttribute(String target, TypedBundle bundle) {
...
     }
 
+    /**
+     * Add a key attribute and the custom variables into the
+     * @param target the id of the target
+     * @param bundle the key attributes bundle containing position etc.
+     * @param custom the customVariables to add at that position
+     */
+    public void addKeyAttribute(String target, TypedBundle bundle, CustomVariable[]custom) {
+        getWidgetState(target, null, 0).setKeyAttribute(bundle,custom);
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:671]: nit: space before custom

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 4]
```diff
@@ -660,6 +661,16 @@ public void addKeyAttribute(String target, TypedBundle bundle) {
         getWidgetState(target, null, 0).setKeyAttribute(bundle);
     }
 
+    /**
+     * Add a key attribute and the custom variables into the
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:665]: into the ???

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 5]
```diff
@@ -898,6 +909,23 @@ public void setKeyAttribute(TypedBundle prop) {
...
+         * Set tge keyAttribute bundle and associated custom attributes
+         * @param prop
+         * @param custom
+         */
+        public void setKeyAttribute(TypedBundle prop, CustomVariable[] custom) {
+            MotionKeyAttributes keyAttributes = new MotionKeyAttributes();
+            prop.applyDelta(keyAttributes);
+            if (custom != null) {
+                for (int i = 0; i < custom.length; i++) {
+                    keyAttributes.mCustom.put( custom[i].getName(), custom[i]);
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:922]: nit: remove space before custom.getName()

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java [hunk 6]
```diff
@@ -285,6 +292,8 @@ private static void parseKeyAttribute(CLObject keyAttribute,
                 false,
         };
         TypedBundle[] bundles = new TypedBundle[frames.size()];
+        CustomVariable[][] customVars  = null;
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:295]: nit: multiple space before =

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java [hunk 7]
```diff
@@ -321,6 +330,71 @@ private static void parseKeyAttribute(CLObject keyAttribute,
                 }
             }
         }
+        // Support for custom attributes in KeyAttributes
+        CLElement customElement = keyAttribute.getOrNull("custom");
+        if (customElement != null && customElement instanceof CLObject) {
+            CLObject customObj = ((CLObject) customElement);
+            int n = customObj.size();
+            customVars =  new CustomVariable[frames.size()][n];
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:338]: nit: multiple space after =

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java [hunk 8]
```diff
@@ -321,6 +330,71 @@ private static void parseKeyAttribute(CLObject keyAttribute,
                 }
             }
         }
+        // Support for custom attributes in KeyAttributes
+        CLElement customElement = keyAttribute.getOrNull("custom");
+        if (customElement != null && customElement instanceof CLObject) {
+            CLObject customObj = ((CLObject) customElement);
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:336]: Not sure if the outside parenthesis is necessary here.
(CLObject) customElement should also work?

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java [hunk 9]
```diff
@@ -321,6 +330,71 @@ private static void parseKeyAttribute(CLObject keyAttribute,
...
+                CLKey key = (CLKey) customObj.get(i);
+                String customName = key.content();
+                Utils.log(" >>>>customObj  "+customObj.content());
+                Utils.log(" >>>>key "+key.content());
+                Utils.log(" >>>>key.value "+key.getValue());
+                Utils.log(" >>>> "+customName);
+
+                if (key.getValue() instanceof CLArray) {
+                    Utils.log(" >>>> "+ customName+" CLArray ");
+                    CLArray arrayValues = (CLArray) key.getValue() ;
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:349]: nit: remove space before ;

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java [hunk 10]
```diff
@@ -321,6 +330,71 @@ private static void parseKeyAttribute(CLObject keyAttribute,
...
+                    int vSize = arrayValues.size();
+                    if (vSize == bundles.length && vSize > 0) {
+                        if (arrayValues.get(0) instanceof CLNumber) {
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:352]: A brief comment would be helpful for understanding what is the if condition and else condition about.

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 11]
```diff
@@ -898,6 +909,23 @@ public void setKeyAttribute(TypedBundle prop) {
...
+                }
+
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:924]: nit: unnecessary newline.

File: projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt [hunk 12]
```diff
@@ -342,3 +346,90 @@ public fun AttributesRotationXY() {
...
+                        }
+                      ]
+                    }
+                  }
+                }
+            }"""),
+            debug = EnumSet.of(MotionLayoutDebugFlags.SHOW_ALL),
+            progress = progress) {
+            val  color = motionProperties("a").value.color("background")
+            var  count = motionProperties("a").value.float("count")
```
Comment (jswong65) [inline projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt:421]: nit: space before count

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java [hunk 13]
```diff
@@ -321,6 +330,71 @@ private static void parseKeyAttribute(CLObject keyAttribute,
...
+                        }
+                    } else {
+                        long cValue = parseColorString(value.content());
+                        Utils.log(" >>>> "+ customName+" color");
+                        if (cValue != -1) {
+                            for (int j = 0; j < bundles.length; j++) {
+                                customVars[j][i] = new CustomVariable(customName,
+                                        TypedValues.Custom.TYPE_COLOR,
+                                        (int) cValue);
+
```
Comment (jswong65) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:390]: nit: unnecessary newline

**Review by jswong65 — APPROVED** [review]
The logic LGTM. Mostly the lint issues.

File: projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt [hunk 14]
```diff
@@ -323,7 +326,8 @@ public fun AttributesRotationXY() {
                           target: ['a'],
                           frames: [25,50,75],
                           rotationX: [0, 45, 60],
-                          rotationY: [60, 45, 0], 
+                          rotationY: [60, 45, 0],
+
```
[the remark this answers is not in the document]
Comment (jafu888) [reply projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt:330]: done

File: projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt [hunk 15]
```diff
@@ -342,3 +346,90 @@ public fun AttributesRotationXY() {
...
+                          
+                        }
+                      ]
+                    }
+                  }
+                }
+            }"""),
+            debug = EnumSet.of(MotionLayoutDebugFlags.SHOW_ALL),
+            progress = progress) {
+            val  color = motionProperties("a").value.color("background")
```
[the remark this answers is not in the document]
Comment (jafu888) [reply projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt:420]: done

File: projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/LaMotion.kt [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+            """)
+
+        MotionLayout(
+            modifier    = Modifier
+                .wrapContentWidth()
+                .align(CenterHorizontally)
+                .height(400.dp)
+                .background(Color.Cyan),
+            motionScene = scene1,
+            progress   = progress.value) {
```
[the remark this answers is not in the document]
Comment (jafu888) [reply projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/LaMotion.kt:98]: done

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java (hunk 9 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:349]: done

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java (hunk 10 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:352]: done

File: projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt (hunk 12 above)
Comment (jafu888) [reply projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/AttributesDemo.kt:421]: done

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java (hunk 13 above)
Comment (jafu888) [reply constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/TransitionParser.java:390]: done

**Pull request merged**